tsl/src/backup.c

```diff
@@ -0,0 +1,192 @@
+#include "catalog.h"
+#include "debug_wait.h"
+#include "dist_util.h"
+#include "remote/dist_commands.h"
+#include "backup.h"
+
+enum
+{
+	Anum_restore_point_node_name = 1,
+	Anum_restore_point_node_type,
+	Anum_restore_point_lsn,
+	_Anum_restore_point_max
+};
+
+static Datum
+create_restore_point_datum(TupleDesc tupdesc, const char *node_name, XLogRecPtr lsn)
+{
+	Datum values[_Anum_restore_point_max] = { 0 };
+	bool nulls[_Anum_restore_point_max] = { false };
+	HeapTuple tuple;
+
+	tupdesc = BlessTupleDesc(tupdesc);
+	if (node_name == NULL)
+	{
+		nulls[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = true;
+		values[AttrNumberGetAttrOffset(Anum_restore_point_node_type)] =
+			CStringGetTextDatum("access_node");
+	}
+	else
+	{
+		values[AttrNumberGetAttrOffset(Anum_restore_point_node_name)] = CStringGetDatum(node_name);
+		values[AttrNumberGetAttrOffset(Anum_restore_point_node_type)] =
+			CStringGetTextDatum("data_node");
+	}
+	values[AttrNumberGetAttrOffset(Anum_restore_point_lsn)] = LSNGetDatum(lsn);
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+	return HeapTupleGetDatum(tuple);
+}
+
+Datum
+create_distributed_restore_point(PG_FUNCTION_ARGS)
+{
+	const char *name = TextDatumGetCString(PG_GETARG_DATUM(0));
+	DistCmdResult *result_cmd;
+	FuncCallContext *funcctx;
+	XLogRecPtr lsn;
+
+	if (SRF_IS_FIRSTCALL())
+	{
+		int name_len = strlen(name);
+		MemoryContext oldctx;
+		TupleDesc tupdesc;
+		char *sql;
+
+		if (name_len >= MAXFNAMELEN)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("restore point name is too long"),
+					 errdetail("Maximum length is %d, while provided name has %d chars.",
+							   MAXFNAMELEN - 1,
+							   name_len)));
+
+		if (RecoveryInProgress())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 (errmsg("recovery is in progress"),
+					  errhint("WAL control functions cannot be executed during recovery."))));
+
+		if (!XLogIsNeeded())
+			ereport(ERROR,
+					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+					 errmsg("WAL level '%s' is not sufficient for creating a restore point",
+							GetConfigOptionByName("wal_level", NULL, false)),
+					 errhint(
+						 "wal_level must be set to \"replica\" or \"logical\" at server start.")));
+
+		if (!superuser())
+			ereport(ERROR,
+					(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+					 errmsg("must be superuser to create restore point")));
+
+		if (!ts_guc_enable_2pc)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("two-phase commit transactions are not enabled"),
+					 errhint("timescale.enable_2pc must be set to true.")));
+
+		if (dist_util_membership() != DIST_MEMBER_ACCESS_NODE)
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_OPERATION_NOT_SUPPORTED),
+					 errmsg("distributed restore point must be created on the access node"),
+					 errhint("pg_create_restore_point() can be used to create restore point "
+							 "on the data node.")));
```

Comment:
I am not sure we need this hint because it can be misleading and it suggests a solution that doesn't achieve the original goal, i.e., to create a distributed restore point. If the user tries to do X we shouldn't assume the user was really trying to do Y and direct them towards doing that instead. Further, a standard restore point on a data node isn't very useful so we should assume that the user simply aimed to create a distributed restore point but executed it on the wrong node. Either remove the hint or change it to something like:

```suggestion
					 errhint("Connect to the access node and create the distributed restore point from there.")));
```